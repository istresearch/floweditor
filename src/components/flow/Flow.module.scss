@import 'variables.module.scss';

.empty_flow {
  margin: 30px;
  width: 400px;

  h1 {
    font-family: 'Roboto', sans-serif;
    letter-spacing: 1px;
    padding: 0;
    text-transform: none;
    font-weight: 100;
    font-size: 18px;
    margin: 0;
  }

  div {
    margin-bottom: 20px;
    font-size: 14px;
    line-height: 18px;
  }
}