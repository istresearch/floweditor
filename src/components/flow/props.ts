import { Type } from 'config/interfaces';
import { AnyAction, ContactProperties, FlowIssue } from 'flowTypes';
import { Asset, AssetStore, AssetType, RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { DispatchWithState, GetState } from 'store/thunks';
import { titleCase } from 'utils';

export interface IssueProps {
  helpArticles: { [key: string]: string };
  issues: FlowIssue[];
}

export interface ActionFormProps extends IssueProps {
  // action details
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;
  assetStore: AssetStore;

  addAsset(assetType: string, asset: Asset): void;

  // update handlers
  updateAction(
    action: AnyAction,
    onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
  ): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface RouterFormProps extends IssueProps {
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;

  assetStore: AssetStore;

  // update handlers
  updateRouter(renderNode: RenderNode): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface LocalizationFormProps extends IssueProps {
  language: Asset;
  nodeSettings: NodeEditorSettings;
  updateLocalizations(languageCode: string, localizations: any[]): void;
  onClose(canceled: boolean): void;
  helpArticles: { [key: string]: string };
}

export const NAME_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Name),
  id: ContactProperties.Name,
  type: AssetType.ContactProperty
};

export const CHANNEL_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Channel),
  id: ContactProperties.Channel,
  type: AssetType.ContactProperty
};

export const LANGUAGE_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Language),
  id: ContactProperties.Language,
  type: AssetType.ContactProperty
};

export const STATUS_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Status),
  id: ContactProperties.Status,
  type: AssetType.ContactProperty
};
