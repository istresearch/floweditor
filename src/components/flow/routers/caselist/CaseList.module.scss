@import 'variables.module.scss';

.case_list_container {
  overflow: hidden;
  border-radius: $curvature;
  margin-top: 8px;
  &:focus {
    outline: none;
  }

  &.scrolling {
    margin-top: 12px;
    background: rgb(250, 250, 250);
  }

  .case_list {
    margin: 4px 0px;
    position: relative;
    overflow-y: auto;
    max-height: 190px;
    padding-bottom: 10px;
    padding-top: 10px;
  }
}

.kase {
  z-index: 2000;

  &:last-child {
    :global {
      .fe-x {
        opacity: 0 !important;
        cursor: default;
      }
    }
    .dnd_icon {
      opacity: 0;
    }  
  }

}
