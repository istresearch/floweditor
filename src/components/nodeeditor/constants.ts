export const DEFAULT_OPERAND = '@input.text';
export const GROUPS_OPERAND = '@contact.groups';
export const WEBHOOK_OPERAND = '@run.webhook';
export const SUBFLOW_OPERAND = '@child.run.status';
export const MEDIA_OPERAND = '@input';
export const DIGITS_OPERAND = '@input.text';
export const MENU_OPERAND = '@input.text';
export const SCHEMES_OPERAND = '@(urn_parts(contact.urn).scheme)';
export const DIAL_OPERAND = '@(default(resume.dial.status, ""))';

// default body for non-GET webhook actions
export const DEFAULT_BODY: string = `@(json(object(
  "contact", object(
    "uuid", contact.uuid, 
    "name", contact.name, 
    "urn", contact.urn
  ),
  "flow", object(
    "uuid", run.flow.uuid, 
    "name", run.flow.name
  ),
  "results", foreach_value(results, extract_object, "value", "category")
)))`;
