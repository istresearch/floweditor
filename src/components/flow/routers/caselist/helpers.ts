import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { Operator } from 'config/interfaces';
import { createUUID } from 'utils';

export const createEmptyCase = (operator: Operator): CaseProps => {
  const uuid = createUUID();
  return {
    uuid,
    kase: {
      uuid,
      type: operator.type,
      arguments: [''],
      category_uuid: null
    },
    categoryName: '',
    valid: true
  };
};
