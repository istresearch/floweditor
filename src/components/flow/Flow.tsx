import { default as bindCallbacks } from 'auto-bind/react';
import Button from 'components/button/Button';
import { Canvas } from 'components/canvas/Canvas';
import { CanvasDraggableProps } from 'components/canvas/CanvasDraggable';
import Node from 'components/flow/node/Node';
import { getDraggedFrom } from 'components/helpers';
import NodeEditor from 'components/nodeeditor/NodeEditor';
import Simulator from 'components/simulator/Simulator';
import Sticky, { STICKY_BODY, STICKY_TITLE } from 'components/sticky/Sticky';
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { FlowDefinition, FlowMetadata, FlowPosition } from 'flowTypes';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Plumber from 'services/Plumber';
import { DragSelection, DebugState } from 'store/editor';
import { RenderNode } from 'store/flowContext';
import { createEmptyNode, detectLoops, getOrderedNodes } from 'store/helpers';
import { NodeEditorSettings } from 'store/nodeEditor';
import AppState from 'store/state';
import {
  ConnectionEvent,
  DispatchWithState,
  mergeEditorState,
  MergeEditorState,
  NoParamsAC,
  onConnectionDrag,
  OnConnectionDrag,
  OnOpenNodeEditor,
  onOpenNodeEditor,
  onRemoveNodes,
  OnRemoveNodes,
  OnUpdateCanvasPositions,
  onUpdateCanvasPositions,
  resetNodeEditingState,
  UpdateConnection,
  updateConnection,
  updateSticky,
  UpdateSticky
} from 'store/thunks';
import {
  createUUID,
  isRealValue,
  NODE_PADDING,
  renderIf,
  snapToGrid,
  timeEnd,
  timeStart
} from 'utils';
import Debug from 'utils/debug';

import styles from './Flow.module.scss';
import { Trans } from 'react-i18next';
import { PopTabType } from 'config/interfaces';
import i18n from 'config/i18n';

declare global {
  interface Window {
    fe: any;
  }
}

export interface FlowStoreProps {
  ghostNode: RenderNode;
  debug: DebugState;
  translating: boolean;
  popped: string;
  dragActive: boolean;

  mergeEditorState: MergeEditorState;

  definition: FlowDefinition;
  nodes: { [uuid: string]: RenderNode };
  metadata: FlowMetadata;
  nodeEditorSettings: NodeEditorSettings;

  updateConnection: UpdateConnection;
  onOpenNodeEditor: OnOpenNodeEditor;
  onUpdateCanvasPositions: OnUpdateCanvasPositions;
  onRemoveNodes: OnRemoveNodes;
  resetNodeEditingState: NoParamsAC;
  onConnectionDrag: OnConnectionDrag;
  updateSticky: UpdateSticky;
}

export interface Translations {
  [uuid: string]: any;
}

export const DRAG_THRESHOLD = 3;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
  return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
  const left = Math.min(drag.startX, drag.currentX);
  const top = Math.min(drag.startY, drag.currentY);
  const width = Math.max(drag.startX, drag.currentX) - left;
  const height = Math.max(drag.startY, drag.currentY) - top;
  return {
    left,
    top,
    width,
    height
  };
};

export class Flow extends React.PureComponent<FlowStoreProps, {}> {
  private Plumber: Plumber;
  private nodeContainerUUID: string;

  // Refs
  private ghost: any;

  public static contextTypes = {
    config: fakePropType
  };

  constructor(props: FlowStoreProps, context: ConfigProviderContext) {
    super(props, context);

    this.nodeContainerUUID = createUUID();

    this.Plumber = new Plumber();

    /* istanbul ignore next */
    if (context.config.debug) {
      window.fe = new Debug(props, this.props.debug);
    }

    bindCallbacks(this, {
      include: [/Ref$/, /^on/, /^is/, /^get/, /^handle/]
    });

    timeStart('Loaded Flow');
  }

  private ghostRef(ref: any): any {
    return (this.ghost = ref);
  }

  public componentDidMount(): void {
    this.Plumber.bind('connection', (event: ConnectionEvent) =>
      this.props.updateConnection(event.sourceId, event.targetId)
    );
    this.Plumber.bind('beforeDrag', (event: ConnectionEvent) => {
      this.beforeConnectionDrag(event);
    });

    this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
      this.props.onConnectionDrag(event, this.context.config.flowType);
    });

    this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
      this.onConnectorDrop(event)
    );
    this.Plumber.bind(
      'beforeStartDetach',
      (event: ConnectionEvent) => !this.props.translating && this.context.config.mutable
    );
    this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
    this.Plumber.bind('beforeDrop', (event: ConnectionEvent) => this.onBeforeConnectorDrop(event));
    this.Plumber.triggerLoaded(this.context.config.onLoad);

    timeEnd('Loaded Flow');
  }

  public componentWillUnmount(): void {
    this.Plumber.reset();
  }

  /**
   * Called right before a connector is dropped onto a new node
   */
  private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
    this.props.resetNodeEditingState();
    const fromNodeUUID = event.sourceId.split(':')[0];
    try {
      detectLoops(this.props.nodes, fromNodeUUID, event.targetId);
    } catch {
      return false;
    }
    return true;
  }

  /**
   * Called the moment a connector is done dragging, whether it is dropped on an
   * existing node or on to empty space.
   */
  private onConnectorDrop(event: ConnectionEvent): boolean {
    const ghostNode = this.props.ghostNode;
    // Don't show the node editor if we a dragging back to where we were
    if (isRealValue(ghostNode) && !isDraggingBack(event)) {
      // Wire up the drag from to our ghost node
      this.Plumber.recalculate(ghostNode.node.uuid);

      const dragPoint = getDraggedFrom(ghostNode);

      this.Plumber.connect(dragPoint.nodeUUID + ':' + dragPoint.exitUUID, ghostNode.node.uuid);

      // Save our position for later
      const { left, top } = (this.ghost &&
        snapToGrid(this.ghost.ele.offsetLeft, this.ghost.ele.offsetTop)) || { left: 0, top: 0 };

      this.props.ghostNode.ui.position = { left, top };

      let originalAction = null;
      if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
        originalAction = ghostNode.node.actions[0];
      }

      // Bring up the node editor
      this.props.onOpenNodeEditor({
        originalNode: ghostNode,
        originalAction
      });
    }

    if (isDraggingBack(event)) {
      this.props.mergeEditorState({ ghostNode: null });
    }

    /* istanbul ignore next */
    document.removeEventListener('mousemove', (window as any).ghostListener);

    return true;
  }

  private beforeConnectionDrag(event: ConnectionEvent): boolean {
    if (event.source) {
      event.source.dispatchEvent(new Event('disconnect'));
    }
    return !this.props.translating;
  }

  private handleStickyCreation(props: CanvasDraggableProps) {
    const stickyMap = this.props.definition._ui.stickies || {};
    const uuid = props.uuid;
    return <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} selected={props.selected} />;
  }

  private handleNodeCreation(props: CanvasDraggableProps) {
    const onlyNode = Object.keys(this.props.nodes).length === 1;
    return (
      <Node
        onlyNode={onlyNode}
        startingNode={props.idx === 0}
        selected={props.selected}
        key={props.uuid}
        data-spec={nodeSpecId}
        nodeUUID={props.uuid}
        plumberMakeTarget={this.Plumber.makeTarget}
        plumberRemove={this.Plumber.remove}
        plumberRecalculate={this.Plumber.recalculate}
        plumberMakeSource={this.Plumber.makeSource}
        plumberConnectExit={this.Plumber.connectExit}
        plumberUpdateClass={this.Plumber.updateClass}
      />
    );
  }

  private getNodes(): CanvasDraggableProps[] {
    return getOrderedNodes(this.props.nodes).map((renderNode: RenderNode, idx: number) => {
      return {
        uuid: renderNode.node.uuid,
        position: renderNode.ui.position,
        elementCreator: this.handleNodeCreation,
        config: renderNode,
        idx
      };
    });
  }

  private getStickies(): CanvasDraggableProps[] {
    const stickyMap = this.props.definition._ui.stickies || {};
    return Object.keys(stickyMap).map((uuid: string, idx: number) => {
      return {
        uuid,
        elementCreator: this.handleStickyCreation,
        position: stickyMap[uuid].position,
        idx
      };
    });
  }

  private getDragNode(): JSX.Element {
    return isRealValue(this.props.ghostNode) ? (
      <div
        data-spec={ghostNodeSpecId}
        key={this.props.ghostNode.node.uuid}
        style={{ position: 'absolute', display: 'block', visibility: 'hidden' }}
      >
        <Node
          onlyNode={false}
          selected={false}
          startingNode={false}
          ref={this.ghostRef}
          ghost={true}
          nodeUUID={this.props.ghostNode.node.uuid}
          plumberMakeTarget={this.Plumber.makeTarget}
          plumberRemove={this.Plumber.remove}
          plumberRecalculate={this.Plumber.recalculate}
          plumberMakeSource={this.Plumber.makeSource}
          plumberConnectExit={this.Plumber.connectExit}
          plumberUpdateClass={this.Plumber.updateClass}
        />
      </div>
    ) : null;
  }

  private getSimulator(): JSX.Element {
    return renderIf(this.context.config.endpoints && this.context.config.endpoints.simulateStart)(
      <Simulator
        key="simulator"
        popped={this.props.popped}
        mergeEditorState={this.props.mergeEditorState}
        onToggled={(visible: boolean, tab: PopTabType) => {
          this.props.mergeEditorState({
            popped: visible ? tab : null
          });
        }}
      />
    );
  }

  private getNodeEditor(): JSX.Element {
    return renderIf(this.props.nodeEditorSettings !== null)(
      <NodeEditor
        key="node-editor"
        helpArticles={this.context.config.help}
        plumberConnectExit={this.Plumber.connectExit}
      />
    );
  }

  // TODO: this should be a callback from the canvas
  private handleDoubleClick(position: FlowPosition): void {
    const { left, top } = position;
    this.props.updateSticky(createUUID(), {
      position: snapToGrid(left - 90 + NODE_PADDING, top - 40),
      title: STICKY_TITLE,
      body: STICKY_BODY
    });
  }

  private getEmptyFlow(): JSX.Element {
    return (
      <div key="create_node" className={styles.empty_flow}>
        <Trans i18nKey="empty_flow_message">
          <h1>Let's get started</h1>
          <div>
            We recommend starting your flow by sending a message. This message will be sent to
            anybody right after they join the flow. This is your chance to send a single message or
            ask them a question.
          </div>
        </Trans>

        <Button
          name={i18n.t('buttons.create_message', 'Create Message')}
          onClick={() => {
            const emptyNode = createEmptyNode(null, null, 1, this.context.config.flowType);
            this.props.onOpenNodeEditor({
              originalNode: emptyNode,
              originalAction: emptyNode.node.actions[0]
            });
          }}
        />
      </div>
    );
  }

  /* 
  public componentDidUpdate(prevProps: FlowStoreProps): void {
    traceUpdate(this, prevProps);
  }
  */

  public handleDragging(uuids: string[]): void {
    uuids.forEach((uuid: string) => {
      try {
        const ele = document.getElementById(uuid);
        const exits = ele.querySelectorAll('.jtk-connected');
        this.Plumber.revalidate([ele, ...exits]);
      } catch (error) {}
    });
  }

  public handleCanvasLoaded(): void {
    this.Plumber.setContainer('canvas');
  }

  public render(): JSX.Element {
    const nodes = this.getNodes();

    const draggables = this.getStickies().concat(nodes);

    return (
      <div>
        {nodes.length === 0 ? this.getEmptyFlow() : <>{this.getSimulator()}</>}
        {this.getNodeEditor()}

        <Canvas
          mutable={this.context.config.mutable}
          draggingNew={!!this.props.ghostNode && !this.props.nodeEditorSettings}
          newDragElement={this.getDragNode()}
          onDragging={this.handleDragging}
          uuid={this.nodeContainerUUID}
          dragActive={this.props.dragActive}
          mergeEditorState={this.props.mergeEditorState}
          onRemoveNodes={this.props.onRemoveNodes}
          draggables={draggables}
          onDoubleClick={this.handleDoubleClick}
          onUpdatePositions={this.props.onUpdateCanvasPositions}
          onLoaded={this.handleCanvasLoaded}
        ></Canvas>
        <div id="activity_recent_contacts"></div>
      </div>
    );
  }
}

/* istanbul ignore next */
const mapStateToProps = ({
  flowContext: { definition, metadata, nodes },
  editorState: { ghostNode, debug, translating, popped, dragActive },
  // tslint:disable-next-line: no-shadowed-variable
  nodeEditor: { settings }
}: AppState) => {
  return {
    nodeEditorSettings: settings,
    definition,
    nodes,
    metadata,
    ghostNode,
    debug,
    translating,
    popped,
    dragActive
  };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
  bindActionCreators(
    {
      mergeEditorState,
      resetNodeEditingState,
      onConnectionDrag,
      onOpenNodeEditor,
      onUpdateCanvasPositions,
      onRemoveNodes,
      updateConnection,
      updateSticky
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Flow);
