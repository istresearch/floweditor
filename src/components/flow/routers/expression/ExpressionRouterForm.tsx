import { default as bindCallbacks } from 'auto-bind/react';
import * as React from 'react';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import CaseList, { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { nodeToState, stateToNode } from 'components/flow/routers/expression/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import { FormState, mergeForm, StringEntry } from 'store/nodeEditor';
import { Alphanumeric, Required, StartIsNonNumeric, validate } from 'store/validators';
import i18n from 'config/i18n';

// TODO: Remove use of Function
// tslint:disable:ban-types
export enum InputToFocus {
  args = 'args',
  min = 'min',
  max = 'max',
  exit = 'exit'
}

export interface ExpressionRouterFormState extends FormState {
  cases: CaseProps[];
  resultName: StringEntry;
  operand: StringEntry;
}

export const leadInSpecId = 'lead-in';

export default class ExpressionRouterForm extends React.Component<
  RouterFormProps,
  ExpressionRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);

    this.state = nodeToState(this.props.nodeSettings);

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  private handleUpdate(keys: {
    resultName?: string;
    operand?: string;
    cases?: CaseProps[];
  }): boolean {
    const updates: Partial<ExpressionRouterFormState> = {};

    if (keys.hasOwnProperty('operand')) {
      updates.operand = validate(i18n.t('forms.operand', 'Operand'), keys.operand, [Required]);
    }

    if (keys.hasOwnProperty('cases')) {
      updates.cases = keys.cases;
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        Alphanumeric,
        StartIsNonNumeric
      ]);
    }

    const updated = mergeForm(this.state, updates);

    // update our form
    this.setState(updated);
    return updated.valid;
  }

  private handleUpdateResultName(resultName: string): void {
    this.handleUpdate({ resultName });
  }

  private handleOperandUpdated(operand: string): void {
    this.handleUpdate({ operand });
  }

  private handleCasesUpdated(cases: CaseProps[]): void {
    this.setState({ cases });
  }

  private handleSave(): void {
    // if we still have invalid cases, don't move forward
    const invalidCase = this.state.cases.find((caseProps: CaseProps) => !caseProps.valid);
    if (invalidCase) {
      return;
    }

    // validate our result name in case they haven't interacted
    const valid = this.handleUpdate({
      resultName: this.state.resultName.value,
      operand: this.state.operand.value
    });

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  public renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <p>If the expression...</p>
        <TextInputElement
          name={i18n.t('forms.operand', 'Operand')}
          showLabel={false}
          autocomplete={true}
          onChange={this.handleOperandUpdated}
          entry={this.state.operand}
        />
        <CaseList
          data-spec="cases"
          cases={this.state.cases}
          onCasesUpdated={this.handleCasesUpdated}
        />
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
