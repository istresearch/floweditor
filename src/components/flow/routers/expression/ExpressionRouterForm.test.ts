import { RouterFormProps } from 'components/flow/props';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import ExpressionRouterForm from 'components/flow/routers/expression/ExpressionRouterForm';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Operators, Types } from 'config/interfaces';
import { RouterTypes, SwitchRouter } from 'flowTypes';
import { composeComponentTestUtils, mock } from 'testUtils';
import { createRenderNode, getRouterFormProps, createMatchRouter } from 'testUtils/assetCreators';
import * as utils from 'utils';
import { createUUID } from 'utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const routerNode = createMatchRouter([]);
routerNode.ui = {
  position: { left: 0, top: 0 },
  type: Types.split_by_expression
};

const { setup } = composeComponentTestUtils<RouterFormProps>(
  ExpressionRouterForm,
  getRouterFormProps(routerNode)
);

describe(ExpressionRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { instance, props } = setup(true, {
        $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
      });

      instance.handleUpdateResultName('Favorite Color');
      instance.handleCasesUpdated([
        {
          kase: { type: Operators.has_any_word, arguments: ['red'] },
          categoryName: 'Red',
          valid: true
        },
        {
          kase: { type: Operators.has_any_word, arguments: ['maroon'] },
          categoryName: 'Red',
          valid: true
        },
        {
          kase: { type: Operators.has_any_word, arguments: ['green'] },
          categoryName: 'Green',
          valid: true
        }
      ] as CaseProps[]);

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).toHaveBeenCalled();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('should cancel', () => {
      const { instance, props } = setup(true, {
        $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
      });

      instance.handleOperandUpdated('@date.now');
      instance.getButtons().secondary.onClick();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled();
    });
  });
});
