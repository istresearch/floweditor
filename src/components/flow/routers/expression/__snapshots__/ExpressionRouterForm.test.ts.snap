// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpressionRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_expression"
  title="Split by Expression"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by a custom expression",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Split by Expression",
        "type": "split_by_expression",
      }
    }
    onChange={[MockFunction]}
  />
  <p>
    If the expression...
  </p>
  <TextInputElement
    autocomplete={true}
    entry={
      Object {
        "value": "@input.text",
      }
    }
    name="Operand"
    onChange={[Function]}
    showLabel={false}
  />
  <CaseList
    cases={Array []}
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText={
      <Trans
        i18nKey="forms.result_name_help"
        values={
          Object {
            "resultFormat": "@results",
          }
        }
      >
        By naming the result, you can reference it later using [[resultFormat]]
      </Trans>
    }
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`ExpressionRouterForm updates should save changes 1`] = `
Object {
  "cases": Array [
    Object {
      "categoryName": "Red",
      "kase": Object {
        "arguments": Array [
          "red",
        ],
        "type": "has_any_word",
      },
      "valid": true,
    },
    Object {
      "categoryName": "Red",
      "kase": Object {
        "arguments": Array [
          "maroon",
        ],
        "type": "has_any_word",
      },
      "valid": true,
    },
    Object {
      "categoryName": "Green",
      "kase": Object {
        "arguments": Array [
          "green",
        ],
        "type": "has_any_word",
      },
      "valid": true,
    },
  ],
  "operand": Object {
    "value": "@input.text",
  },
  "resultName": Object {
    "validationFailures": Array [],
    "value": "Favorite Color",
  },
  "valid": true,
}
`;

exports[`ExpressionRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        },
        Object {
          "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
        },
        Object {
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "red",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "maroon",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "green",
            ],
            "category_uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
            "type": "has_any_word",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "name": "Red",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
          Object {
            "exit_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "name": "Green",
            "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
          },
          Object {
            "exit_uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
            "name": "Other",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
        ],
        "default_category_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        "operand": "@input.text",
        "result_name": "Favorite Color",
        "type": "switch",
      },
      "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
    },
    "ui": Object {
      "config": Object {
        "cases": Object {},
      },
      "position": null,
      "type": "split_by_expression",
    },
  },
]
`;
